gnome-image-installer/pages/confirm/gis-confirm-page.c

```diff
@@ -0,0 +1,281 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
+ *
+ */
+
+/* Confirm page {{{1 */
+
+#define PAGE_ID "confirm"
+
+#include "config.h"
+#include "confirm-resources.h"
+#include "gis-confirm-page.h"
+#include "gis-store.h"
+
+#include <udisks/udisks.h>
+#include <glib/gstdio.h>
+#include <glib/gi18n.h>
+#include <gio/gio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+struct _GisConfirmPagePrivate {
+    GtkLabel *vendor_label;
+    GtkLabel *product_label;
+
+    GtkLabel *image1_label;
+    GtkLabel *model1_label;
+    GtkLabel *device1_label;
+    GtkLabel *size1_label;
+
+    GtkBox *warning_box;
+    GtkLabel *warning_label;
+
+    GtkBox *wrong_computer_box;
+
+    GtkLabel *countdown_label;
+    GtkButton *cancel_button;
+    GtkButton *continue_button;
+
+    guint countdown_source;
+    guint countdown_remaining;
+};
+typedef struct _GisConfirmPagePrivate GisConfirmPagePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (GisConfirmPage, gis_confirm_page, GIS_TYPE_PAGE);
+G_DEFINE_QUARK(confirm-error, gis_confirm_error);
+
+static void
+gis_confirm_page_advance (GisConfirmPage *self)
+{
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+  GisPage *page = GIS_PAGE (self);
+
+  if (priv->countdown_source != 0)
+    {
+      g_source_remove (priv->countdown_source);
+      priv->countdown_source = 0;
+    }
+
+  gis_assistant_next_page (gis_driver_get_assistant (page->driver));
+}
+
+static gboolean
+gis_confirm_page_countdown_cb (gpointer data)
+{
+  GisConfirmPage *self = GIS_CONFIRM_PAGE (data);
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+
+  if (priv->countdown_remaining == 0)
+    {
+      gis_confirm_page_advance (self);
+      return G_SOURCE_REMOVE;
+    }
+  else
+    {
+      g_autofree gchar *countdown_message =
+        g_strdup_printf (_("Installation will commence automatically in %d seconds."),
+                         priv->countdown_remaining);
+      gtk_label_set_text (priv->countdown_label, countdown_message);
+      priv->countdown_remaining--;
+      return G_SOURCE_CONTINUE;
+    }
+}
+
+static void
+gis_confirm_page_start_countdown (GisConfirmPage *self)
+{
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+
+  priv->countdown_remaining = 30;
```

Comment:
How is this chosen? Might a comment briefly explaining why 30 is the magic number be a good idea?